### PR TITLE
Bste/input args

### PR DESCRIPTION
**Pull request by benjstephenson**

### CODE REVIEW

File: lib/types.ts [hunk 1]
```diff
@@ -9,6 +13,7 @@ export interface NumericalBase extends OperationBase {
 }
 
 export interface AddOperation extends NumericalBase {
```
Comment (benjstephenson) [inline lib/types.ts:15]: we only allow a single input here, is that what we want?

File: lib/types.ts [hunk 2]
```diff
@@ -46,30 +51,30 @@ export interface OrOperation extends OperationBase {
...
 export interface RootOperation extends NumericalBase {
-	index: number | NumericOperation
+	index: FromParamOperation | NumericOperation
 	operation: "root"
-	radicand: number | NumericOperation
+	radicand: FromParamOperation | NumericOperation
 	returns: "number"
 }
 
 export interface SubtractOperation extends NumericalBase {
```
Comment (benjstephenson) [inline lib/types.ts:67]: We take the same input for both sides of the operation here, which would evaluate down to 0.  Should we restrict it to one side only?

File: lib/types.ts [hunk 3]
```diff
@@ -21,20 +26,20 @@ export interface AndOperation extends OperationBase {
 }
 
 export interface DivideOperation extends NumericalBase {
```
Comment (benjstephenson) [inline lib/types.ts:28]: Ignore most of these changes, there only there to make sure the compile fails so I don't forget to update them

File: lib/mathematical/subtract/index.ts [hunk 4]
```diff
@@ -1,40 +1,46 @@
...
+import isFromParamOperation from "../../utilities/isFromParamOperation"
+import { identity } from "@sitebender/fp/lib/functions"
 
 export type SubtractF = (
 	o: SubtractOperation,
-) => () => Either<Array<string>, Option<number>>
+) => (input: Option<number>) => Either<Array<string>, Option<number>>
+
+const divide: SubtractF = op => input => {
+	const evaluate = (
```
Comment (benjstephenson) [inline lib/mathematical/subtract/index.ts:20]: I should be able to lift this out to somewhere else

File: lib/mathematical/subtract/index.ts [hunk 5]
```diff
@@ -1,40 +1,46 @@
...
-					pipe(
-						[minuend, subtrahend],
-						sequence,
-						map(([minuend, subtrahend]) => {
-							const difference = minuend - subtrahend
+		pipe([evaluate(op.minuend), evaluate(op.subtrahend)], e.allOf(identity)),
+		e.flatMap(([minued, subtrahend]) =>
+			pipe(
+				[minued, subtrahend],
+				o.sequence,
```
Comment (benjstephenson) [inline lib/mathematical/subtract/index.ts:34]: if either value is missing, we stop.  Is that the behaviour we want everywhere?

File: lib/mathematical/subtract/index.ts (hunk 4 above)
Comment (chasm) [reply lib/mathematical/subtract/index.ts:20]: Shouldn't this be called `subtract` not `divide`?

File: lib/types.ts (hunk 1 above)
Comment (chasm) [reply lib/types.ts:15]: Let's chat about this tomorrow. I'm becoming more convinced that we should wrap constants:

```ts
// Should these extend NumericOperation?
export interface IntegerOperation {
  operand: number
  operation: "integer"
  returns: "integer"
}

export interface PrecisionOperation {
  operand: number
  operation: "precision"
  precision: number
  returns: "precision"
}

export interface StringOperation {
  operand: string
  operation: "string"
  returns: "string"
}
```

Something along those lines. Then the `operands` or equivalent are _all_ Operations and we just create functions that return the operand.

With regard to the AddOperation, then, maybe:

```ts
export interface AddOperation {
  addends: Array<NumericOperation | FromParamOperation>
  operation: "add"
  returns: "number"
}
```

Off the top of my head ...

File: lib/types.ts (hunk 2 above)
Comment (chasm) [reply lib/types.ts:67]: This is an interesting situation. How would we narrow it?

File: lib/mathematical/subtract/index.ts (hunk 5 above)
Comment (chasm) [reply lib/mathematical/subtract/index.ts:34]: I would think so. How else could it work? There is no point in continuing down one side if the other is undefined, no?

File: lib/mathematical/subtract/index.ts (hunk 4 above)
Comment (benjstephenson) [reply lib/mathematical/subtract/index.ts:20]: most probably 😅

File: lib/mathematical/subtract/index.ts (hunk 5 above)
Comment (benjstephenson) [reply lib/mathematical/subtract/index.ts:34]: I agree; I tried out using a default value in the add operation but I think maybe it's better to just not perform the op if all the inputs aren't available the way we do here.

File: lib/mathematical/multiply/index.ts [hunk 6]
```diff
@@ -1,31 +1,24 @@
 import type { MultiplyOperation } from "../../types"
 
-import { Option, some } from "@sitebender/fp/lib/option"
-import { Either, right } from "@sitebender/fp/lib/either"
+import { Option } from "@sitebender/fp/lib/option"
+import { Either } from "@sitebender/fp/lib/either"
 import { map } from "@sitebender/fp/lib/array"
 import { default as multiplyArray } from "@sitebender/fp/lib/array/reduce/multiply"
```
Comment (chasm) [inline lib/mathematical/multiply/index.ts:6]: I know this isn't a change, but if we're going to reach down into the `array` folder for this, why not just reach one more into `array/map`? I prefer when accessing these to access them as the default whenever possible. And to be consistent.

I just cleaned up the FP module and have been meaning to tidy up this one, too.

**Review by chasm — APPROVED** [review]
This looks great. I can't wait to try the "constants" idea (everything is an operation).

File: lib/mathematical/multiply/index.ts (hunk 6 above)
Comment (benjstephenson) [reply lib/mathematical/multiply/index.ts:6]: Ok, I'll tidy that import style up as I go then.

Comment (chasm) [reply lib/mathematical/multiply/index.ts:6]: I can do it if you get bored.

**Pull request merged**